### PR TITLE
Add TCM access control page

### LINKED ISSUE #3634: [TCM] Role-based access control
**Product:** Tarantool Cluster Manager
**Since:** TCM 1.0.0, Tarantool 3.0
**Audience/target:** admins
**Root document:** TBD
**SME:** @ filonenko-mikhail 

# Details

- users
- passwords/secrets
- roles
- permissions: administrative and cluster permissions
- sessions
- password policy

### PR DESCRIPTION
**Pull request by p7nov**
Resolves #3634

Deployment: https://docs.d.tarantool.io/en/doc/gh-3634-tcm-rbac/reference/tooling/tcm/tcm_access_control/

Add new section [**TCM > Access Control**](https://docs.d.tarantool.io/en/doc/gh-3634-tcm-rbac/reference/tooling/tcm/tcm_access_control/):
- Permissions: administrative and cluster, concepts
- Roles: concepts, management instructions
- Users: concepts, management instructions
- Passwords/secrets: definitions and concepts, management, expiry, blocking, password policy
- Sessions
- Permissions reference

TODO (in scope of #3637 ):

- Move permissions reference to future section **TCM > Reference**
- Move password policy items description to  future section **TCM > Reference**.

### CODE REVIEW

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+    :header-rows: 1
+
+    *   -   Permission
+        -   Description
+
+    *   -   ``cluster.lowlevel.state.read``
+        -   Read low-level information about cluster configuration (for debug purposes)
+
+    *   -   ``cluster.lowlevel.state.write``
+        -   Write low-level information about cluster configuration (for debug purposes)
```
Comment (filonenko-mikhail) [inline doc/reference/tooling/tcm/tcm_access_control.rst]: move it the bottom please to make it

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        -   View roles' details
+
+    *   -   ``admin.roles.write``
+        -   Edit roles and add new roles
+
+    *   -   ``admin.lowlevel.state.read``
+        -   Read low-level information from |tcm| storage (for debug purposes)
+
+    *   -   ``admin.lowlevel.state.write``
+        -   Write low-level information to |tcm| storage (for debug purposes)
```
Comment (filonenko-mikhail) [inline doc/reference/tooling/tcm/tcm_access_control.rst]: move it bottom please

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        -   View password policy
+
+    *   -   ``admin.passwordpolicy.write``
+        -   Manage password policy
+
+    *   -  ``admin.devmode.toggle``
+        -   Toggle development mode
+
+    *   -   ``user.password.change``
+        -   Change own password
```
Comment (filonenko-mikhail) [inline doc/reference/tooling/tcm/tcm_access_control.rst]: let's remove. seems that we dont use it in tcm, sorry

**Review by filonenko-mikhail — APPROVED** [review]

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+|tcm_full_name| features a role-based access control system. It enables flexible
+management of access to |tcm| functions, connected clusters, and stored data.
+There are three main entities in the |tcm| access system: permissions, roles,
+and users (or user accounts). They work as follows:
+
+-   Permissions correspond to specific functions or objects in
+    |tcm| (*administrative permissions*) or operations on clusters (*cluster permissions*).
+-   Roles are predefined sets of *administrative* permissions to
+    assign to users.
+-   Users have roles that define their access rights to |tcm| function and objects, and
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]:
```suggestion
-   Users have roles that define their access rights to |tcm| functions and objects, and
```

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+    *cluster* permissions that are assigned for each cluster separately.
+
+..  note::
+
+    |tcm| users, roles, and permissions are not to be confused with similar subjects
+    of :ref:`Tarantool access control system <authentication>`. To access Tarantool
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]:
```suggestion
    of the :ref:`Tarantool access control system <authentication>`. To access Tarantool
```

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+    define pages shown in the left menu. *Write* permissions define the availability
+    of controls for managing objects on the pages.
+    For example, users with read permission to clusters can view the **Clusters** page
+    but they don't see **Add**, **Edit**, or **Remove** buttons unless they have the
+    write permission.
+
+    Administrative permissions are assigned to users through :ref:`roles <tcm_access_control_roles>`.
+
+*   *Cluster permissions* enable actions with connected Tarantool clusters.
+    These permission are granted to users on a per-cluster level: each user has a separate
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]:
```suggestion
    These permissions are granted to users on a per-cluster level: each user has a separate
```

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+    set of permissions for each cluster.
+
+    Technically, cluster permissions define pages shown in the **Cluster** section
+    of the left menu and controls available on these pages. For example, users
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]: Don't quite understand to which word `controls` refers: permissions, section, menu?

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+Administrators can create new users, edit, and delete existing ones.
+
+The tools for managing users are located on the **Users** page.
+
+To create a user:
+
+1.  Click **Add**.
+2.  Fill in the user information: username, full name, and description.
+3.  Generate or enter a password.
+4.  Select roles to assign to user.
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]:
```suggestion
4.  Select roles to assign to the user.
```

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+|tcm| uses the general term *secret* for user authentication keys. A secret is any
+pair of a public and a private key that can be used for authentication. In |tcm| |tcm_version|,
+*password* is the only supported secret type. In this case, the public key is a username,
+and the private key is a password.
+
+Users receive their first passwords during the account creation. It can be entered
+manually or generated automatically.
+
+All passwords are governed by the :ref:`password policy <tcm_access_control_password_policy>`.
+It can be flexibly configured to follow security requirements of your organization.
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]:
```suggestion
It can be flexibly configured to follow the security requirements of your organization.
```

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+
+To change a user's password, click **Edit** in the **Actions** menu of the corresponding
+**Secrets** table row and enter the new password in the **New secret key** field.
+
+Password expiry
+~~~~~~~~~~~~~~~
+
+Passwords expire automatically after the expiration period defined in the :ref:`password policy <tcm_access_control_password_policy>`.
+When a user logs in to |tcm| with an expired password, the only action available to
+them is password change. All other |tcm| functions and objects are unavailable until
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]:
```suggestion
them is a password change. All other |tcm| functions and objects are unavailable until
```

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+the new password is set.
+
+Administrators can also set users' password to expired manually.
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]:
```suggestion
Administrators can also set users' passwords to expired manually.
```

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+
+..  admonition:: Enterprise Edition
+    :class: fact
+
+    |tcm_full_name| is a part of the `Enterprise Edition <https://www.tarantool.io/compare/>`_.
+
+|tcm_full_name| features a role-based access control system. It enables flexible
+management of access to |tcm| functions, connected clusters, and stored data.
+There are three main entities in the |tcm| access system: permissions, roles,
+and users (or user accounts). They work as follows:
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst:14]: > They work as follows:

Too many nouns in the previous sentence to which `they` can refer. As an idea, we can remove this part completely:

```
permissions, roles, and users (or user accounts). 
They work as follows:
```

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+..  note::
+
+    You can delete a role only if there are no users with this role.
+
+Users
+-----
+
+|tcm| users gain access to objects and actions through assigned :ref:`roles <tcm_access_control_roles>`
+and :ref:`cluster permissions <tcm_access_control_permissions>`.
+Each user can have any number of roles, including zero. Users without roles
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]: > including zero

this sounds OK for API docs but for Web UI sounds a bit weird

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+
+Administrators can create new users, edit, and delete existing ones.
+
+The tools for managing users are located on the **Users** page.
+
+To create a user:
+
+1.  Click **Add**.
+2.  Fill in the user information: username, full name, and description.
+3.  Generate or enter a password.
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]: Not sure but to me `enter a password` sounds like it should be entered when logging in. I'd think about replacing it with `specify`.

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+Passwords
+---------
+
+|tcm| uses the general term *secret* for user authentication keys. A secret is any
+pair of a public and a private key that can be used for authentication. In |tcm| |tcm_version|,
+*password* is the only supported secret type. In this case, the public key is a username,
+and the private key is a password.
+
+Users receive their first passwords during the account creation. It can be entered
+manually or generated automatically.
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]: I'd remove the second sentence here as it duplicates the info from `Managing users` (and added the link for `account creation` to  `Managing users`).

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+Changing your password
+~~~~~~~~~~~~~~~~~~~~~~
+
+To change your own password, click your name in the top-right corner and go to
+**Settings** > **Change password**.
+
+Changing users' passwords
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Administrators can manage a user's passwords on this user's **Secrets** page.
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]: > a user's passwords

sounds like a user can have multiple passwords

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+
+-   **Minimal password length**.
+-   **Do not use last N passwords**.
+-   **Password expiration in days**. After this number of days, the user loses access
+    to any objects and functions except password change. After they change the password,
+    all the access rights are returned.
+-   **Password expiration warning in days**. After this number of days, the user
+    sees a warning that they must change their password.
+-   **Block after N login attempts**.
+-   **User lockout time in seconds**.
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]: Are these settings related? Don't quite understand the second one

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+        -   Revoke users' sessions
+
+    *   -   ``admin.ldap.read``
+        -   View LDAP configurations
+
+    *   -   ``admin.ldap.write``
+        -   Manage LDAP configurations
+
+    *   -   ``admin.passwordpolicy.read``
+        -   View password policy
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst:325]: Looks like the first column should be a bit wider:
image

**Review by andreyaksenov — CHANGES_REQUESTED** [review]

File: doc/reference/tooling/tcm/tcm_access_control.rst (hunk 16 above)
Comment (filonenko-mikhail) [reply doc/reference/tooling/tcm/tcm_access_control.rst]: it's true

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+Administrative permissions
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following administrative permissions are available in |tcm|:
+
+..  list-table::
+    :widths: 30 70
+    :header-rows: 1
+
+    *   -   Permission
```
Comment (filonenko-mikhail) [inline doc/reference/tooling/tcm/tcm_access_control.rst:276]: my bad, 

user.password.change is actual for access to change own password, please restore it

File: doc/reference/tooling/tcm/tcm_access_control.rst (hunk 18 above)
Comment (p7nov) [reply doc/reference/tooling/tcm/tcm_access_control.rst:325]: Fixed

File: doc/reference/tooling/tcm/tcm_access_control.rst (hunk 19 above)
Comment (p7nov) [reply doc/reference/tooling/tcm/tcm_access_control.rst:276]: Restored

File: doc/reference/tooling/tcm/tcm_access_control.rst (hunk 17 above)
Comment (p7nov) [reply doc/reference/tooling/tcm/tcm_access_control.rst]: Added the descriptions

File: doc/reference/tooling/tcm/tcm_access_control.rst (hunk 16 above)
Comment (p7nov) [reply doc/reference/tooling/tcm/tcm_access_control.rst]: This subject is confusing me a bit. The case of one TCM user having multiple secrets (username/password pairs) is technically possible. I didn't describe it on purpose: I don't see a valid real-life usage scenario for this. 

AFAIU, this functionality is intended for use when TCM receives more than one way to authenticate users. For example, one user will have three secrets: a password, a certificate, and a fingerprint (whatever, just imagining :)). But right now it doesn't look useful on practice.

@filonenko-mikhail, what do you think? should we describe the case of multiple passwords explicitly (therefore encouraging readers to do this)?

**Comment by filonenko-mikhail** [comment]
No we should not
Lets describe that password is one

Ср, 22 нояб. 2023 г. в 06:41, Pavel Semyonov ***@***.***>:

File: doc/reference/tooling/tcm/tcm_access_control.rst [hunk 20]
```diff
@@ -0,0 +1,393 @@
...
+
+-   **Minimal password length**.
+-   **Do not use last N passwords**.
+-   **Password expiration in days**. Users' passwords :ref:`expire <tcm_access_control_password_expiry>`
+    after this number of days since they were set. Users with expired passwords
+    lose access to any objects and functions except password change until they set
+    a new password.
+-   **Password expiration warning in days**. After this number of days, the user
+    sees a warning that their password expires soon.
+-   **Block after N login attempts**. Temporarily block user if they enter their
```
Comment (andreyaksenov) [inline doc/reference/tooling/tcm/tcm_access_control.rst]:
```suggestion
-   **Block after N login attempts**. Temporarily block users if they enter their
```

**Review by andreyaksenov — APPROVED** [review]
